source/Timer.cpp

```diff
@@ -0,0 +1,376 @@
+			out.Write("idle", sqrt(idleMaxSpeed));
+		if(requirePeaceful)
+			out.Write("peaceful");
+		if(optional)
+			out.Write("optional");
+		if(requireUncloaked)
+			out.Write("uncloaked");
+		if(repeatReset)
+			out.Write("repeat reset");
+		if(resetFired)
+			out.Write("reset fired");
+		if(resetCondition != ResetCondition::PAUSE)
+		{
+			static const map<ResetCondition, string> resets = {
+				{Timer::ResetCondition::NONE, "none"},
+				{Timer::ResetCondition::LEAVE_SYSTEM, "leave system"},
+				{Timer::ResetCondition::LEAVE_ZONE, "leave zone"}
+			};
+			auto it = resets.find(resetCondition);
+			if(it != resets.end())
+				out.Write("reset", it->second);
+		}
+		if(proximity > 0.)
+		{
+			string distance = "close";
+			if(!closeTo)
+				distance = "far";
+			if(proximityCenter)
+				out.Write("proximity", proximityCenter->Name());
+			else if(!proximityCenters.IsEmpty())
+			{
+				out.Write("proximity");
+				proximityCenters.Save(out);
+			}
+			out.Write("proximity settings", proximity, distance);
+		}
+
+		for(const auto &it : actions)
+			it.second.Save(out);
+	}
+	out.EndChild();
+}
+
+
+
+// Calculate the total time to wait, including any random value,
+// and instantiate the triggered action.
+Timer Timer::Instantiate(map<string, string> &subs, const System *origin, int jumps, int64_t payload) const
+{
+	Timer result;
+	result.requireIdle = requireIdle;
+	result.requireUncloaked = requireUncloaked;
+	result.system = system;
+	result.systems = systems;
+	result.proximity = proximity;
+	result.proximityCenter = proximityCenter;
+	result.proximityCenters = proximityCenters;
+	result.closeTo = closeTo;
+	result.resetCondition = resetCondition;
+	result.repeatReset = repeatReset;
+	result.resetFired = resetFired;
+	result.idleMaxSpeed = idleMaxSpeed;
+	result.requirePeaceful = requirePeaceful;
+
+	// Validate all the actions attached to the timer, and if they're *all* valid, instantiate them too
+	string reason;
+	auto ait = actions.begin();
+	for( ; ait != actions.end(); ++ait)
+	{
+		reason = ait->second.Validate();
+		if(!reason.empty())
+			break;
+	}
+	if(ait != actions.end())
+	{
+		Logger::LogError("Instantiation Error: Timer action \"" + TriggerToText(ait->first)
+			+ "\" uses invalid " + std::move(reason));
+		return result;
+	}
+	for(const auto &it : actions)
+		result.actions[it.first] = it.second.Instantiate(subs, origin, jumps, payload);
+
+	// Calculate the random variance to the wait time.
+	result.waitTime = waitTime;
+	if(randomWaitTime > 1)
+		result.waitTime += Random::Int(randomWaitTime);
+
+	// We also build a cache of the matching proximity object(s) for the instantiated Timer.
+	// This avoids having to do all these comparisons every Step().
+	for(const StellarObject &proximityObject : system->Objects())
+		if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
+			proximityCenters.Matches(proximityObject.GetPlanet())))
+			result.proximityCache.insert(&proximityObject);
```

Comment:
Nullptr exception if there is no system. Also, we only need to build a cache if we actually have a proximity center defined.
```suggestion
	if(system && (proximityCenter || !proximityCenters.isEmpty()))
		for(const StellarObject &proximityObject : system->Objects())
			if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
				proximityCenters.Matches(proximityObject.GetPlanet())))
				result.proximityCache.insert(&proximityObject);
```